src/tiledb/cloud/geospatial/ingestion.py

```diff
@@ -0,0 +1,1287 @@
+
+            logger.info("max memory usage: %.3f GiB", max_memory_usage() / (1 << 30))
+
+
+def ingest_raster_udf(
+    chunks: Optional[Tuple[GeoBlockMetadata, ...]] = None,
+    *,
+    dataset_uri: str,
+    extents: Optional[BoundingBox] = None,
+    band_count: Optional[int] = None,
+    dtype: Optional[str] = None,
+    res: Optional[Tuple[float, float]] = None,
+    crs: Optional[str] = None,
+    nodata: Optional[float] = None,
+    tile_size: int = RASTER_TILE_SIZE,
+    resampling: str = "bilinear",
+    append: bool = False,
+    verbose: bool = False,
+    config: Optional[Mapping[str, Any]] = None,
+    compressor: Optional[dict] = None,
+    id: str = "raster",
+    trace: bool = False,
+    log_uri: Optional[str] = None,
+):
+    """Internal udf that ingests server side batch of raster files
+    into tiledb arrays using Rasterio API
+
+    :param dataset_uri: str, output TileDB array name
+    :param chunks: tuple, sequence of GeoBlockMetadata objects containing
+        the destination raster window and the input files
+        that contribute to this window
+    :param extents: Extents of the destination raster
+    :param band_count: int, number of bands in destination array
+    :param dtype: str, dtype of destination array
+    :param res: Tuple[float, float], resolution in x/y of the destination raster
+    :param crs: str, CRS for the destination dataset
+    :param nodata: float, NODATA value for destination raster
+    :param resampling: string, resampling method,
+        one of None, bilinear, cubic, nearest and average
+    :param append: bool, whether to append to the array
+    :param verbose: verbose logging, defaults to False
+    :param config: dict, configuration to pass on tiledb.VFS
+    :param compressor: dict, serialized compression filter
+    :param id: str, ID for logging, defaults to None
+    :param trace, bool, enable trace logging
+    :param log_uri: log array URI
+    """
+
+    import rasterio
+    import rasterio.merge
+
+    import tiledb
+    from tiledb.cloud.utilities import Profiler
+    from tiledb.cloud.utilities import get_logger_wrapper
+    from tiledb.cloud.utilities import max_memory_usage
+
+    logger = get_logger_wrapper(verbose)
+
+    with tiledb.scope_ctx(config):
+        with Profiler(array_uri=log_uri, id=id, trace=trace):
+            vfs = tiledb.VFS()
+
+            if not append:
+                import rasterio
+                from rasterio.transform import from_origin
+
+                from tiledb import filter
+
+                if (
+                    extents is None
+                    or res is None
+                    or crs is None
+                    or band_count is None
+                    or dtype is None
+                ):
+                    raise ValueError(
+                        """CRS, band count, extents and resolution are
+                        required to create empty raster array"""
```

Comment:
To do nice strings when it's longer than will fit in one line, you can use the fact that adjacent literals are automatically concatenated:

```python
raise ValueError(
    "CRS, band count, ..."
    " required to ..."
)
```

That means you end up with one string that looks like "CRS, band count, ... required to ..." without excessive white space.

Reply:
👍